File: web-app/frontend/lib/api/residence.ts
/**
 * Server-only fetch helper for the resident view (/residence).
 *
 * The resident is NOT a NextAuth B2B user, so this sends NO bearer token. The
 * resident's identity is passed as the `X-Resident-Id` header; the backend only
 * honors it when RESIDENT_DEV_MODE=1 (P4-1 dev override), and returns 401
 * otherwise. P4-3 (magic-link session) will replace the header source with a real
 * resident session — this fetcher's shape stays the same.
 *
 * Types must stay in lockstep with backend app/schemas/residence.py.
 */

export type ResidenceKpi = {
  eui_kwh_m2_yr: number | null
  eui_climate_adjusted_kwh_m2_yr: number | null
  climate_adjustment_factor: number | null
  epc_band: string | null
  heating_dhw_kwh_annual: number | null
  building_avg_eui_kwh_m2_yr: number | null
  vs_building_pct: number | null
  coverage_start: string | null
  coverage_end: string | null
  cov_days: number | null
}

export type ResidenceCommonArea = {
  unit_metered_kwh: number | null
  unit_allocated_kwh: number | null
  allocation_share: number | null
  cons_weight: number | null
  area_weight: number | null
  cons_share: number | null
  area_share: number | null
  building_total_kwh: number | null
  coverage_start: string | null
  coverage_end: string | null
}

export type ResidenceMonthlyPoint = {
  year: number
  month: number
  energy_type: string
  kwh: number | null
  cost_eur: number | null
  building_avg_kwh: number | null
  vs_building_pct: number | null
}

export type ResidenceUnit = {
  unit_id: string
  area_m2: number | null
  is_heated: boolean | null
  kpi: ResidenceKpi | null
  common_area: ResidenceCommonArea | null
  monthly: ResidenceMonthlyPoint[]
}

export type ResidenceSummary = {
  as_of: string
  units: ResidenceUnit[]
}

type FetchResult<T> = { ok: true; data: T } | { ok: false; error: string }

function getBackendUrl(): string {
  // Defaults to the local uvicorn instance; production overrides via env.
  return process.env.BACKEND_URL || "http://127.0.0.1:8000"
}

/**
 * Server-side fetch of the resident's own-unit summary.
 *
 * `residentId` is the resident_identity UUID (dev: from the ?resident= query
 * param; later: from the resident session). No auth bearer — see module docstring.
 */
export type ResidentIdentity = { token?: string | null; devResidentId?: string | null }

export async function fetchResidenceSummary(
  identity: ResidentIdentity
): Promise<FetchResult<ResidenceSummary>> {
  const url = `${getBackendUrl()}/residence/summary`
  const headers: Record<string, string> = { Accept: "application/json" }
  if (identity.token) {
    headers["Authorization"] = `Bearer ${identity.token}`
  } else if (identity.devResidentId) {
    headers["X-Resident-Id"] = identity.devResidentId
  }
  try {
    const res = await fetch(url, {
      method: "GET",
      headers,
      cache: "no-store",
    })

    if (!res.ok) {
      const body = await res.text().catch(() => "<no body>")
      if (res.status === 503 && body.includes("fabric_unavailable")) {
        return { ok: false, error: "fabric_unavailable" }
      }
      if (res.status === 401) {
        // Dev override off, or no/invalid resident id.
        return { ok: false, error: "resident_unauthorized" }
      }
      console.error(`[fetchResidenceSummary] Backend ${res.status}: ${body.slice(0, 200)}`)
      return { ok: false, error: `Backend ${res.status}` }
    }

    const data = (await res.json()) as ResidenceSummary
    return { ok: true, data }
  } catch (err) {
    console.error("[fetchResidenceSummary] Network/unknown error:", err)
    return { ok: false, error: "Network error" }
  }
}

File: web-app/backend/app/schemas/residence.py
"""Pydantic response models for the /residence (resident-view) endpoints.

The resident sees ONLY their own unit(s): per-unit heating/DHW EUI + EPC band, the
HKVO common-area allocation, and the monthly UVI series. The building benchmark is
already anonymized + pre-aggregated INTO each unit's own Gold row
(``building_avg_*`` + ``vs_building_pct``) -- so these models never carry another
named unit's data. See ``docs/architecture/residential-data-model.md`` (privacy
posture) and ``unified-access-model.md`` (resident scope).
"""
from datetime import date

from pydantic import BaseModel, Field


class ResidenceKpi(BaseModel):
    """Per-unit headline KPIs (from ``gold_residential_unit_kpi``)."""

    eui_kwh_m2_yr: float | None = Field(
        default=None, description="Heating+DHW energy use intensity, annualized."
    )
    eui_climate_adjusted_kwh_m2_yr: float | None = Field(
        default=None,
        description="HDD-normalized EUI. Equals raw EUI while climate_adjustment_factor=1.0.",
    )
    climate_adjustment_factor: float | None = Field(
        default=None,
        description="HDD normalization factor. 1.0 = not yet wired (no weather for the building).",
    )
    epc_band: str | None = Field(
        default=None, description="A-E band on the RESIDENTIAL_MF scale (50/80/120/180)."
    )
    heating_dhw_kwh_annual: float | None = Field(
        default=None, description="Annualized heating+DHW consumption (kWh)."
    )
    building_avg_eui_kwh_m2_yr: float | None = Field(
        default=None, description="Anonymized building-average EUI (benchmark)."
    )
    vs_building_pct: float | None = Field(
        default=None, description="This unit's EUI vs the building average, percent."
    )
    coverage_start: date | None = None
    coverage_end: date | None = None
    cov_days: int | None = Field(
        default=None, description="Days of metered coverage behind the annualization."
    )


class ResidenceCommonArea(BaseModel):
    """The unit's share of building energy under the HKVO split (from
    ``gold_residential_common_split``). Default weighting is 70% consumption /
    30% area (``cons_weight`` / ``area_weight``)."""

    unit_metered_kwh: float | None = None
    unit_allocated_kwh: float | None = Field(
        default=None, description="kWh allocated to this unit after the HKVO split."
    )
    allocation_share: float | None = Field(
        default=None, description="This unit's fraction of the building total (0..1)."
    )
    cons_weight: float | None = Field(default=None, description="HKVO consumption weight (default 0.70).")
    area_weight: float | None = Field(default=None, description="HKVO area weight (default 0.30).")
    cons_share: float | None = None
    area_share: float | None = None
    building_total_kwh: float | None = None
    coverage_start: date | None = None
    coverage_end: date | None = None


class ResidenceMonthlyPoint(BaseModel):
    """One UVI series point: a (year, month, energy_type) row from
    ``gold_residential_uvi_monthly``. This is the data behind the legally-required
    monthly consumption information (EED/HKVO)."""

    year: int
    month: int
    energy_type: str = Field(description="'heating' or 'dhw'.")
    kwh: float | None = None
    cost_eur: float | None = None
    building_avg_kwh: float | None = Field(
        default=None, description="Anonymized building-average for this month/type."
    )
    vs_building_pct: float | None = None


class ResidenceUnit(BaseModel):
    """Everything the resident sees for one of their units."""

    unit_id: str
    area_m2: float | None = None
    is_heated: bool | None = None
    kpi: ResidenceKpi | None = Field(
        default=None, description="None when the unit has no Gold KPI row yet."
    )
    common_area: ResidenceCommonArea | None = None
    monthly: list[ResidenceMonthlyPoint] = Field(default_factory=list)


class ResidenceSummary(BaseModel):
    """The /residence/summary payload.

    A resident usually occupies one unit, but ``resolve_resident_scope`` returns a
    set (a resident could hold more than one tenancy), so this is a list.
    """

    as_of: date = Field(description="Date the tenancy-window scope was resolved for.")
    units: list[ResidenceUnit] = Field(default_factory=list)


class ResidentConsumeRequest(BaseModel):
    """Body for POST /residence/auth/consume — the raw magic-link token."""

    token: str = Field(description="The raw token from the magic-link URL.")


class ResidentSession(BaseModel):
    """Resident session JWT returned by /residence/auth/consume."""

    resident_token: str
    expires_in_hours: int
